### PR TITLE
Add optional reward scaling

### PR DESCRIPTION
**Pull request by reciprocated**
In the spirit of #48
reward scaling from https://github.com/DLR-RM/stable-baselines3/blob/d5d1a02c15cdce868c72bbc94913e66fdd2efd3a/stable_baselines3/common/vec_env/vec_normalize.py#L220
minibatch whitening from https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/pytorch/ppo/ppo.py#L80

https://wandb.ai/sorry/public/reports/mean_reward-22-11-17-01-59-30---VmlldzoyOTg0ODc3

### CODE REVIEW

File: configs/ppo_config.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ model:
 
 train:
   seq_length: 48  # Size of LM context
-  epochs: 1000 # Train for max(epochs, total_steps)
```
Comment (LouisCastricato) [inline configs/ppo_config.yml:9]: Thank god we finally decreased this haha

File: trlx/orchestrator/ppo_orchestrator.py [hunk 2]
```diff
@@ -66,14 +71,34 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):
...
             response_tensors = samples[:, query_tensors.shape[1] :]
             texts = self.rl_model.tokenizer.batch_decode(
                 samples, skip_special_tokens=True
             )
-            scores = torch.as_tensor(self.score(texts))
+            exp_score_time = time()
+            scores = torch.as_tensor(self.score(texts), device=samples.device)
+            stats["exp_score_time"] = time() - exp_score_time
+
+            if self.ref_mean is None:
```
Comment (LouisCastricato) [inline trlx/orchestrator/ppo_orchestrator.py:87]: Some comments here about what this does would be helpful :)

File: trlx/utils/modeling.py [hunk 3]
```diff
@@ -34,3 +55,49 @@ def flatten_dict(
...
+        if dist.is_initialized():
+            xs_mean, xs_var, xs_count = get_global_statistics(xs)
+        else:
+            xs_count = xs.numel()
+            xs_var, xs_mean = torch.var_mean(xs, unbiased=False)
+
+        delta = xs_mean - self.mean
+        tot_count = self.count + xs_count
+
+        m_a = self.var * self.count
```
Comment (LouisCastricato) [inline trlx/utils/modeling.py:94]: I hate having lots of math and no intuitive explanation of what the math is doing as comments. Please fix.

**Review by LouisCastricato — COMMENTED** [review]
Looks good. Just needs more comments.

**Comment by LouisCastricato** [comment]
Can you post the W&B link to a run where we can confirm rescaling works? Thanks.

**Comment by Dahoas** [comment]
> Can you post the W&B link to a run where we can confirm rescaling works? Thanks.

Agreed let's make sure to do this on all algo changing PRs.

File: configs/ppo_config.yml [hunk 4]
```diff
@@ -35,6 +35,8 @@ method:
   cliprange: 0.2  # clip range
   cliprange_value: 0.2  # clip range
   vf_coef: 2.3  # value term weight
+  scale_reward: True
+  clip_reward: 10
```
Comment (Dahoas) [inline configs/ppo_config.yml:39]: Shouldn't reward generally be in the range [-1,1]?

File: trlx/orchestrator/ppo_orchestrator.py [hunk 5]
```diff
@@ -66,14 +71,34 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):
...
+                self.ref_mean, self.ref_std = scores.mean(), scores.std()
```
Comment (Dahoas) [inline trlx/orchestrator/ppo_orchestrator.py:88]: Perhaps naming this ref_mean is a bit misleading? It is not the mean of the reference model but rather the mean of the training model.

**Comment by LouisCastricato** [comment]
Run looks good to me. Add comments and we can merge!

File: trlx/utils/modeling.py [hunk 6]
```diff
@@ -2,12 +2,33 @@
...
+    global_mean = global_sum / count
+
+    sum_var = torch.sum((xs - global_mean) ** 2)
+    dist.all_reduce(sum_var, dist.ReduceOp.SUM)
+    global_var = sum_var / count
+    return global_mean, global_var, count
+
+
+def whiten(xs: torch.Tensor, shift_mean=True, distributed=True) -> torch.Tensor:
+    """Whitens values"""
```
Comment (Dahoas) [inline trlx/utils/modeling.py:25]: Out of curiosity do we have a reference for whitening? (Some blog post, arxiv paper)

File: trlx/utils/modeling.py [hunk 7]
```diff
@@ -34,3 +55,49 @@ def flatten_dict(
...
+        {
+            f"{name}/mean": mean,
+            f"{name}/min": torch.where(mask.bool(), xs, np.inf).min(),
+            f"{name}/max": torch.where(mask.bool(), xs, -np.inf).max(),
+            f"{name}/std": torch.sqrt(((xs - mean) * mask).pow(2).sum() / n),
+        }
+    )
+
+
+class RunningMoments:
```
Comment (Dahoas) [inline trlx/utils/modeling.py:72]: Very nice

File: trlx/utils/modeling.py [hunk 8]
```diff
@@ -34,3 +55,49 @@ def flatten_dict(
...
+    def __init__(self):
```
Comment (Dahoas) [inline trlx/utils/modeling.py:73]: If we precompute the mean and var of our initial reward distribution ahead of time do we have a way of incorporating that?

**Review by Dahoas — APPROVED** [review]

**Comment by Dahoas** [comment]
Approved though I'd like to know if pre-computed mean, var of baseline rewards can be used as well
